doc/neps/nep-0050-scalar-promotion.rst

```diff
@@ -0,0 +1,729 @@
+of Python scalars and NumPy scalars/0-D arrays is, again, two-fold:
+
+1. The special handling of NumPy scalars/0-D arrays as well as the value
+   inspection can be very surprising to users,
+2. The value-inspection logic is much harder to explain and implement.
+   It is further harder to make it available to user defined DTypes through
+   :ref:`NEP 42 <NEP42>`.
+   Currently, this leads to a dual implementation of a new and an old (value
+   sensitive) system.  Fixing this will greatly simplify the internal logic
+   and make results more consistent.
+
+We believe that the proposal of "weak" Python scalars will help users by
+providing a clearer mental model for which datatype an operation will
+result in.
+This model fits well with the preservation of array precisions that NumPy
+currently often has, and also aggressively does for in-place operations::
+
+    arr += value
+
+Preserves precision so long "kind" boundaries are not crossed.
+
+And while some users will probably miss the value inspecting behavior even for
+those cases where it seems useful, it quickly leads to surprises.  This may be
+expected::
+
+    np.array([100], dtype=np.uint8) + 1000 == np.array([1100], dtype=np.uint16)
+
+But the following will then be a surprise::
+
+    np.array([100], dtype=np.uint8) + 200 == np.array([44], dtype=np.uint8)
+
+Considering this, we believe that the proposal follows the "principle of least
+surprise".
+
+
+Usage and Impact
+================
+
+There will be no transition period due to the difficulty and noise this is
+expected to create.  In rare cases users may need to adjust code to avoid
+reduced precision or incorrect results.
+
+We plan to provide an *optional* warning mode capable of notifying users of
+potential changes in behavior in most relevant cases.
+
+
+Impact on ``can_cast``
+----------------------
+
+Can cast will never inspect the value anymore.  So that the following results
+are expected to change from ``True`` to ``False``::
+
+  np.can_cast(100, np.uint8)
+  np.can_cast(np.int64(100), np.uint8)
+  np.can_cast(np.array(100, dtype=np.int64), np.uint8)
+
+We expect that the impact of this change will be small compared to that of
+the following changes.
+
+.. note::
+
+    The first example where the input is a Python scalar could be preserved
+    to some degree, but this is not currently planned.
+
+
+Impact on operators and functions involving NumPy arrays or scalars
+-------------------------------------------------------------------
+
+The main impact on operations not involving Python scalars (``float``, ``int``,
+``complex``) will be that 0-D arrays and NumPy scalars will never behave
+value-sensitive.
+This removes currently surprising cases.  For example::
+
+  np.arange(10, dtype=np.uint8) + np.int64(1)
+  # and:
+  np.add(np.arange(10, dtype=np.uint8), np.int64(1))
+
+Will return an int64 array because the type of ``np.int64(1)`` is strictly
+honoured.
```

Comment:
It's not clear from context ("currently surprising cases") if this "will" is describing current or future behaviour.